### DISCUSSION TITLE
Ruby 2.4.2 fails to build on Ubuntu 16.04

### DISCUSSION THREAD

**Discussion by kuroda** [rbenv/ruby-build#2126] [Build failures]
I have installed Ruby 2.4.1 successfully on my Ubuntu 16.04 Desktop, but failed to build Ruby 2.4.2.

Here are the last lines of error log:

```
make[2]: Entering directory '/tmp/ruby-build.20171106150424.32381/ruby-2.4.2/ext/digest/md5'
compiling md5init.c
linking shared-object digest/md5.so
/usr/bin/ld: /usr/local/lib/libcrypto.a(md5_dgst.o): relocation R_X86_64_PC32 against symbol `md5_block_asm_data_order' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
Makefile:262: recipe for target '../../../.ext/x86_64-linux/digest/md5.so' failed
make[2]: *** [../../../.ext/x86_64-linux/digest/md5.so] Error 1
make[2]: Leaving directory '/tmp/ruby-build.20171106150424.32381/ruby-2.4.2/ext/digest/md5'
exts.mk:198: recipe for target 'ext/digest/md5/all' failed
make[1]: *** [ext/digest/md5/all] Error 2
make[1]: Leaving directory '/tmp/ruby-build.20171106150424.32381/ruby-2.4.2'
uncommon.mk:220: recipe for target 'build-ext' failed
make: *** [build-ext] Error 2
```

**Comment by Lewington-pitsos** [comment]
I have the same problem on ubuntu 17.10 for all ruby versions I have tried to install

Re-installing rbenv did not seem to help.

**Comment by kuroda** [comment]
[UPDATE]

I succeeded to install Ruby 2.4.2 on the *clean installed* Ubuntu 16.04 Desktop.

It seems that something goes wrong on my *working* Ubuntu 16.04.

I will close this issue after researching just a little bit more.

**Comment by kuroda** [comment]
I have not detected the root cause of the problem, I found a workaround:

1. Build `openssl` from the source:

```
$ cd /tmp
$ curl -LO https://www.openssl.org/source/openssl-1.1.0f.tar.gz
$ tar xvfz openssl-1.1.0f.tar.gz
$ cd openssl-1.1.0f
$ ./config --prefix=/opt/openssl/1.1.0f
$ make
$ sudo make install
```

2. Build ruby with `CONFIGURE_OPTS` environment variable:

```
$ CONFIGURE_OPTS="--with-openssl-dir=/opt/openssl/1.1.0f" rbenv install 2.4.2
```

[UPDATE]

You also need to install `cacert.pem` in order to make SSL connections:

```
$ sudo wget -O /opt/openssl/1.1.0f/ssl/cert.pem http://curl.haxx.se/ca/cacert.pem
```

**Comment by casperl** [comment]
I have the same problem.  Apt-Get installed versiion on Ubuntu 17.10 is Ruby Build version: 20161225-1
No version of Ruby installs now through rbenv.  Error reportd:
BUILD FAILED (Ubuntu 17.10 using ruby-build 20170914-5-gf41e847) .... ends in "segmentation fault, core dumped."
Since rbenv appears to have a gem version of ruby-build (did "gem update" under rbenv do this, because I did not?) I issued "gem uninstall ruby-build) and ran "rbenv install 2.3.1" using the Ubuntu installed version of Ruby-Build, but the same error and version number of Ruby-build appears.

Next step to is to uninstall + purge rbenv and to sea if Ubuntu supplied Ruby will be sufficient for my needs.

**Comment by casperl** [comment]
Purging rbenv and the ~/.rbenv directory and a clean install of rbenv performed.
Running "rbenv install 2.3.1" produced a similar error: 
![image](https://user-images.githubusercontent.com/1303185/32947783-6f296976-cba5-11e7-9632-8d6491435c30.png)

**Comment by kuroda** [comment]
@casperl I think your problem is different from mine.

I don't encounter with any segmentation faults. My problem is related to linker.

**Comment by kuroda** [comment]
@casperl

According to [a bug report](https://bugs.ruby-lang.org/issues/14076), you shoud type:

```
$ sudo apt-get install gcc-6 g++-6
$ CC=/usr/bin/gcc-6 rbenv install 2.3.1
```

**Comment by m-rwash** [comment]
@kuroda that worked for me. Thanks!

**Comment by kuroda** [comment]
Upon the release of Ruby 2.5.0, I tried to build on my Ubuntu 16.04. I got following error messages:

```
BUILD FAILED (Ubuntu 16.04 using ruby-build 20171215-3-ga97a138)

Inspect or clean up the working tree at /tmp/ruby-build.20171226020853.23177
Results logged to /tmp/ruby-build.20171226020853.23177.log

Last 10 log lines:
make[2]: Leaving directory '/tmp/ruby-build.20171226020853.23177/ruby-2.5.0/ext/digest/sha1'
exts.mk:259: recipe for target 'ext/digest/sha1/all' failed
make[1]: *** [ext/digest/sha1/all] Error 2
linking shared-object digest.so
make[2]: Leaving directory '/tmp/ruby-build.20171226020853.23177/ruby-2.5.0/ext/continuation'
make[2]: Leaving directory '/tmp/ruby-build.20171226020853.23177/ruby-2.5.0/ext/-test-/string'
make[2]: Leaving directory '/tmp/ruby-build.20171226020853.23177/ruby-2.5.0/ext/digest'
make[1]: Leaving directory '/tmp/ruby-build.20171226020853.23177/ruby-2.5.0'
uncommon.mk:240: recipe for target 'build-ext' failed
make: *** [build-ext] Error 2
```

As I did for Ruby 2.4.2, after building the `openssl` from the source, I succeeded to compile ruby with this command:

```
CONFIGURE_OPTS="--with-openssl-dir=/opt/openssl/1.1.0f" rbenv install 2.5.0
```

**Comment by mrvijaycode** [comment]
I am also getting the similar error, please could you help me.

![ruby error](https://user-images.githubusercontent.com/1425672/39635819-31584ca0-4fdc-11e8-919b-080543acadca.png)

**Comment by goermine** [comment]
Did you try to install this (below) list of dependency to your Ubuntu system ?:
```sudo apt-get install autoconf bison build-essential libssl-dev libyaml-dev libreadline6-dev zlib1g-dev libncurses5-dev libffi-dev libgdbm3 libgdbm-de```
For me these stuff have fixed issue.

**Comment by kubakrzempek** [comment]
As far as I know, (and tested) Ruby versions < 2.4 requires libssl1.0, while >2.4 libssl1.1+. The two libssl packages conflict with each other, so you can't have both of them, so I had to juggle the libs in order to install the required ruby version. To make things even funnier (or more complicated), Ruby <2.4 tends to require `gcc-6` instead of the current upstream version (which is 7).
So, usually it looks like that:
```
# for ruby <2.4
apt install libssl1.0-dev
CC=$(which gcc-6) rbenv install `version`

# for ruby >2.4
apt install libssl-dev
rbenv install `version`
```

**Comment by RaminMammadzada** [comment]
I also faced this problem. This resource helped me to solve the problem.
https://www.digitalocean.com/community/tutorials/how-to-install-ruby-on-rails-with-rvm-on-ubuntu-18-04

**Comment by MatzFan** [comment]
Can this issue not be closed, as both the `gcc-6` and `libssl` dependencies for earlier Rubies are documented in the wiki ([here](https://github.com/rbenv/ruby-build/wiki#gcc-compatibility) and [here](https://github.com/rbenv/ruby-build/wiki#openssl-usrincludeopensslasn1_mach102-error-error-this-file-is-obsolete-please-update-your-software))?

After installing `gcc-6` and `libssl1.0-dev` I was able to install both pre and post 2.4 Rubies on Ubuntu 18.04 as per the solution offered by @kubakrzempek without a problem. BTW it appears I do have both `libssl` libraries installed:
```
$ /sbin/ldconfig -pv|grep libssl
   libssl3.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libssl3.so
   libssl.so.1.1 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libssl.so.1.1
   libssl.so.1.0.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.0
   libssl.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libssl.so
```

**Comment by robclancy** [comment]
Why should it be closed? What's the point of using a tool that works with all ruby envs if you have to dive into the docs for a compile error? At the very least it should check for these dependencies before building and get them or tell you they are missing so you can install them.

**Comment by jmig5776** [comment]
@kubakrzempek 
> 
> ```
> # for ruby <2.4
> apt install libssl1.0-dev
> CC=$(which gcc-6) rbenv install `version`
> 
> # for ruby >2.4
> apt install libssl-dev
> rbenv install `version`
> ```

this worked very well.

**Comment by bcawrse** [comment]
> ```
> # for ruby <2.4
> apt install libssl1.0-dev
> CC=$(which gcc-6) rbenv install `version`
> 
> # for ruby >2.4
> apt install libssl-dev
> rbenv install `version`
> ```

saved me installing an older Ruby version for a project I inherited.

**Comment by patrickemuller** [comment]
> As far as I know, (and tested) Ruby versions < 2.4 requires libssl1.0, while >2.4 libssl1.1+. The two libssl packages conflict with each other, so you can't have both of them, so I had to juggle the libs in order to install the required ruby version. To make things even funnier (or more complicated), Ruby <2.4 tends to require `gcc-6` instead of the current upstream version (which is 7).
> So, usually it looks like that:
> 
> ```
> # for ruby <2.4
> apt install libssl1.0-dev
> CC=$(which gcc-6) rbenv install `version`
> 
> # for ruby >2.4
> apt install libssl-dev
> rbenv install `version`
> ```

You're an angel!

**Comment by hsbt** [comment]
This issue was fixed at the first reporter.

**Comment by AhlamHani** [comment]
any other solutions? all of them didn't work for me, I'm using ruby-install 2.6.3

```
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/stringio'
compiling stringio.c
stringio.c: In function 'strio_ungetbyte':
stringio.c:814:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
         VALUE v = rb_funcall(c, rb_intern("modulo"), 1, INT2FIX(256));
         ^~~~~
stringio.c:822:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
     const char *cp = RSTRING_PTR(c);
     ^~~~~
stringio.c: At top level:
cc1: warning: unrecognized command line option '-Wno-self-assign'
cc1: warning: unrecognized command line option '-Wno-parentheses-equality'
cc1: warning: unrecognized command line option '-Wno-constant-logical-operand'
cc1: warning: unrecognized command line option '-Wno-cast-function-type'
linking shared-object stringio.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/stringio'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/rbconfig/sizeof'
compiling sizes.c
compiling limits.c
linking shared-object rbconfig/sizeof.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/rbconfig/sizeof'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/bigdecimal'
compiling bigdecimal.c
linking shared-object bigdecimal.so
installing default bigdecimal libraries
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/bigdecimal'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/bigdecimal/util'
compiling util.c
linking shared-object bigdecimal/util.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/bigdecimal/util'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/json'
installing default libraries
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/json'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/json/generator'
compiling generator.c
linking shared-object json/ext/generator.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/json/generator'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/json/parser'
compiling parser.c
linking shared-object json/ext/parser.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/json/parser'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/fiber'
compiling fiber.c
linking shared-object fiber.so
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/fiber'
make[2]: Entering directory '/home/ahlam/src/ruby-2.6.3/ext/ripper'
extracting ripper.y from ../.././parse.y
compiling compiler ripper.y
ripper.y:761.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
  761 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
ripper.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
generating eventids1.c from ../.././parse.y
generating eventids2table.c from ../.././ext/ripper/eventids2.c
compiling ripper.c
ripper.c: In function 'ripper_yyparse':
ripper.c:5400:12: error: 'YYEMPTY' undeclared (first use in this function)
   yychar = YYEMPTY; /* Cause a token to be read.  */
            ^~~~~~~
ripper.c:5400:12: note: each undeclared identifier is reported only once for each function it appears in
ripper.c:5531:22: error: 'YYerror' undeclared (first use in this function)
   if (yychar <= END_OF_INPUT)
                      ^~~~~~~
ripper.c:5537:16: error: 'YYUNDEF' undeclared (first use in this function)
   else if (yychar == YYerror)
                ^~~~~~~
ripper.c:10702:51: error: macro "yydestruct" passed 5 arguments, but takes just 3
           if (yychar == END_OF_INPUT)
                                                   ^
ripper.c:10701:11: error: 'yydestruct' undeclared (first use in this function)
           /* Return failure if at end of input.  */
           ^~~~~~~~~~
ripper.c:10758:65: error: macro "yydestruct" passed 5 arguments, but takes just 3
       /* Pop the current state because it cannot handle the error token.  */
                                                                 ^
ripper.c:10814:47: error: macro "yydestruct" passed 5 arguments, but takes just 3
   if (yychar != YYEMPTY)
                                               ^
ripper.c:10823:65: error: macro "yydestruct" passed 5 arguments, but takes just 3
      this YYABORT or YYACCEPT.  */
                                                                 ^
At top level:
ripper.y:9127:1: warning: 'parser_token_value_print' defined but not used [-Wunused-function]
 parser_token_value_print(struct parser_params *p, enum yytokentype type, const YYSTYPE *valp)
 ^~~~~~~~~~~~~~~~~~~~~~~~
ripper.c:5303:1: warning: 'ruby_parser_yydestruct' defined but not used [-Wunused-function]
 ruby_parser_yydestruct (const char *yymsg,
 ^~~~~~~~~~~~~~~~~~~~~~
cc1: warning: unrecognized command line option '-Wno-self-assign'
cc1: warning: unrecognized command line option '-Wno-parentheses-equality'
cc1: warning: unrecognized command line option '-Wno-constant-logical-operand'
cc1: warning: unrecognized command line option '-Wno-cast-function-type'
make[2]: *** [Makefile:287: ripper.o] Error 1
make[2]: Leaving directory '/home/ahlam/src/ruby-2.6.3/ext/ripper'
make[1]: *** [exts.mk:254: ext/ripper/all] Error 2
make[1]: Leaving directory '/home/ahlam/src/ruby-2.6.3'
make: *** [uncommon.mk:286: build-ext] Error 2
!!! Compiling ruby 2.6.3 failed!

```

**Comment by Nakilon** [comment]
I've installed the ruby 2.6.6 with help of this blog post on how to compile old openssl separately https://deanpcmad.com/2024/installing-older-ruby-versions-on-ubuntu-24-04-and-22-04/ found via reddit. Since it has no comments functionality I'll leave this here.

**ISSUE**

Even though it `rbenv install`s, it failed when I used it:

```none
$ ruby -rdigest/sha1 -e ""
Traceback (most recent call last):
	1: from /home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
/home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': libcrypto.so.1.1: cannot open shared object file: No such file or directory - /home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/x86_64-linux/digest/sha1.so (LoadError)
```
```none
$ file /home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/x86_64-linux/digest/sha1.so
/home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/x86_64-linux/digest/sha1.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=6c08bb3c4e9af9a24f5f4accbfd460f5ae427496, with debug_info, not stripped
```

The `ldd` hinted that it can't see the `libssl.so`:

````none
$ ldd /home/.../.rbenv/versions/2.6.6/lib/ruby/2.6.0/x86_64-linux/openssl.so
	linux-vdso.so.1 (0x00007fff1b1fb000)
	/usr/lib/x86_64-linux-gnu/libscela${PLATFORM}.so.6 => /usr/lib/x86_64-linux-gnu/libscelahaswell.so.6 (0x00007f9a06800000)
	libruby.so.2.6 => /home/.../.rbenv/versions/2.6.6/lib/libruby.so.2.6 (0x00007f9a06400000)
	libssl.so.1.1 => not found
	libcrypto.so.1.1 => not found
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9a06000000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9a06bfd000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f9a06bf8000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f9a06bdc000)
	libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f9a06ba2000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9a06abb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9a06c75000)
````

First I tried to reinstall the ruby with `LDFLAGS=-L$HOME/_/openssl/installed/lib` compilation flag after which the `ldd` sees everything and ruby does not crash. But then `bundle install` fails and the `curl -Lks 'https://git.io/rg-ssl' | ruby` shows all red.

**CAUSE**

After a while I've found that after I followed that blog post few days ago I've moved the "installed openssl" directory to another place but the library appeared to be more dynamic than I though. It has the path hardcoded all over the place.

S**OLUTION**

To recompile the openssl after you move it, or move it back, or symlink to old path like I did.
